File: crates/hyle/src/field.rs
use indexmap::IndexMap;
use serde::{Deserialize, Serialize};
use serde_json::Value as JsonValue;

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum SortType {
    String,
    Numeric,
    Date,
    None,
}

impl Default for SortType {
    fn default() -> Self {
        Self::String
    }
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum Primitive {
    String,
    Number,
    Boolean,
    File,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Reference {
    pub entity: String,
    #[serde(default = "default_display_field")]
    pub display_field: String,
}

fn default_display_field() -> String {
    "name".to_owned()
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", tag = "kind")]
pub enum FieldType {
    Primitive {
        primitive: Primitive,
    },
    Reference {
        reference: Reference,
    },
    Array {
        item: Box<FieldType>,
    },
    Shape {
        fields: IndexMap<String, ShapeField>,
    },
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ShapeField {
    pub label: String,
    #[serde(rename = "type")]
    pub field_type: FieldType,
    #[serde(default)]
    pub options: FieldOptions,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct InputHint {
    pub kind: String,
    #[serde(default, skip_serializing_if = "IndexMap::is_empty")]
    pub props: IndexMap<String, JsonValue>,
}

impl InputHint {
    pub fn new(kind: impl Into<String>) -> Self {
        Self {
            kind: kind.into(),
            props: IndexMap::new(),
        }
    }

    pub fn with_prop(mut self, key: impl Into<String>, value: impl Into<JsonValue>) -> Self {
        self.props.insert(key.into(), value.into());
        self
    }
}

#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FieldOptions {
    #[serde(default)]
    pub sort: SortType,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub input: Option<InputHint>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub fixed_value: Option<JsonValue>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub rule: Option<String>,
    #[serde(default, skip_serializing_if = "IndexMap::is_empty")]
    pub metadata: IndexMap<String, JsonValue>,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Field {
    pub label: String,
    #[serde(rename = "type")]
    pub field_type: FieldType,
    #[serde(default)]
    pub options: FieldOptions,
}

impl Field {
    pub fn new(label: impl Into<String>, field_type: FieldType) -> Self {
        Self {
            label: label.into(),
            field_type,
            options: FieldOptions::default(),
        }
    }

    pub fn string(label: impl Into<String>) -> Self {
        Self::new(
            label,
            FieldType::Primitive {
                primitive: Primitive::String,
            },
        )
        .with_input(InputHint::new("text"))
    }

    pub fn number(label: impl Into<String>) -> Self {
        Self::new(
            label,
            FieldType::Primitive {
                primitive: Primitive::Number,
            },
        )
        .with_sort(SortType::Numeric)
        .with_input(InputHint::new("number"))
    }

    pub fn boolean(label: impl Into<String>) -> Self {
        Self::new(
            label,
            FieldType::Primitive {
                primitive: Primitive::Boolean,
            },
        )
        .with_sort(SortType::None)
        .with_input(InputHint::new("checkbox"))
    }

    pub fn file(label: impl Into<String>) -> Self {
        Self::new(
            label,
            FieldType::Primitive {
                primitive: Primitive::File,
            },
        )
        .with_sort(SortType::None)
        .with_input(InputHint::new("file"))
    }

    pub fn reference(label: impl Into<String>, entity: impl Into<String>) -> Self {
        Self::new(
            label,
            FieldType::Reference {
                reference: Reference {
                    entity: entity.into(),
                    display_field: default_display_field(),
                },
            },
        )
        .with_input(InputHint::new("select"))
    }

    pub fn array(label: impl Into<String>, item: FieldType) -> Self {
        Self::new(
            label,
            FieldType::Array {
                item: Box::new(item),
            },
        )
        .with_sort(SortType::None)
    }

    pub fn shape(label: impl Into<String>, fields: IndexMap<String, ShapeField>) -> Self {
        Self::new(label, FieldType::Shape { fields }).with_sort(SortType::None)
    }

    pub fn with_sort(mut self, sort: SortType) -> Self {
        self.options.sort = sort;
        self
    }

    pub fn with_input(mut self, input: InputHint) -> Self {
        self.options.input = Some(input);
        self
    }

    pub fn with_metadata(mut self, key: impl Into<String>, value: JsonValue) -> Self {
        self.options.metadata.insert(key.into(), value);
        self
    }
}

impl ShapeField {
    pub fn new(label: impl Into<String>, field_type: FieldType) -> Self {
        Self {
            label: label.into(),
            field_type,
            options: FieldOptions::default(),
        }
    }
}

/// Build a `Field` by kind name. Used by the WASM `make_field` export so
/// the JS `field` builder helpers can delegate entirely to Rust.
///
/// `kind`: `"string"` | `"number"` | `"boolean"` | `"file"` | `"ref"`
/// `entity`: required when `kind == "ref"`, ignored otherwise.
pub fn make_field(kind: &str, label: &str, entity: Option<&str>) -> Field {
    match kind {
        "number" => Field::number(label),
        "boolean" => Field::boolean(label),
        "file" => Field::file(label),
        "ref" => Field::reference(label, entity.unwrap_or("")),
        _ => Field::string(label),
    }
}

File: zig-bindings/src/field.zig
const std = @import("std");
const raw = @import("raw.zig");
const Value = raw.Value;
const StringHashMap = raw.StringHashMap;

pub const SortType = enum {
    String,
    Numeric,
    Date,
    None,
};

pub const Primitive = enum {
    String,
    Number,
    Boolean,
    File,
};

pub const Reference = struct {
    entity: []const u8,
    display_field: []const u8,
};

pub const FieldType = union(enum) {
    Primitive: struct { primitive: Primitive },
    Reference: struct { reference: Reference },
    Array: struct { item: *const FieldType },
    Shape: struct { fields: StringHashMap(ShapeField) },
};

pub const ShapeField = struct {
    label: []const u8,
    field_type: FieldType,
    options: FieldOptions,
};

pub const InputHint = struct {
    kind: []const u8,
    props: StringHashMap(Value),

    pub fn new(kind: []const u8) InputHint {
        return .{ .kind = kind, .props = StringHashMap(Value).init(std.testing.allocator) };
    }

    pub fn withProp(self: *InputHint, key: []const u8, value: Value) void {
        self.props.put(key, value) catch {};
    }
};

pub const FieldOptions = struct {
    sort: SortType = .String,
    input: ?InputHint = null,
    fixed_value: ?Value = null,
    rule: ?[]const u8 = null,
    metadata: StringHashMap(Value) = undefined,
};

pub const Field = struct {
    label: []const u8,
    field_type: FieldType,
    options: FieldOptions,

    pub fn new(label: []const u8, field_type: FieldType) Field {
        return .{
            .label = label,
            .field_type = field_type,
            .options = FieldOptions{ .metadata = StringHashMap(Value).init(std.testing.allocator) },
        };
    }

    pub fn string(label: []const u8) Field {
        var f = Field.new(label, .{ .Primitive = .{ .primitive = .String } });
        f.options.input = InputHint.new("text");
        return f;
    }

    pub fn textarea(label: []const u8, rows: u32) Field {
        var f = Field.new(label, .{ .Primitive = .{ .primitive = .String } });
        var hint = InputHint.new("textarea");
        hint.props.put("rows", Value{ .Int = rows }) catch {};
        f.options.input = hint;
        return f;
    }

    pub fn number(label: []const u8) Field {
        var f = Field.new(label, .{ .Primitive = .{ .primitive = .Number } });
        f.options.sort = .Numeric;
        f.options.input = InputHint.new("number");
        return f;
    }

    pub fn boolean(label: []const u8) Field {
        var f = Field.new(label, .{ .Primitive = .{ .primitive = .Boolean } });
        f.options.sort = .None;
        f.options.input = InputHint.new("checkbox");
        return f;
    }

    pub fn file(label: []const u8) Field {
        var f = Field.new(label, .{ .Primitive = .{ .primitive = .File } });
        f.options.sort = .None;
        f.options.input = InputHint.new("file");
        return f;
    }

    pub fn reference(label: []const u8, entity: []const u8) Field {
        var f = Field.new(label, .{ .Reference = .{ .reference = .{ .entity = entity, .display_field = "name" } } });
        f.options.input = InputHint.new("select");
        return f;
    }

    pub fn array(label: []const u8, _item: FieldType) Field {
        _ = _item;
        var f = Field.new(label, .{ .Array = .{ .item = undefined } });
        f.options.sort = .None;
        return f;
    }

    pub fn shape(label: []const u8, fields: StringHashMap(ShapeField)) Field {
        var f = Field.new(label, .{ .Shape = .{ .fields = fields } });
        f.options.sort = .None;
        return f;
    }

    pub fn withSort(self: *Field, sort: SortType) void {
        self.options.sort = sort;
    }

    pub fn withInput(self: *Field, input: InputHint) void {
        self.options.input = input;
    }
};

pub fn makeField(kind: []const u8, label: []const u8, entity: ?[]const u8) Field {
    if (std.mem.eql(u8, kind, "number")) return Field.number(label);
    if (std.mem.eql(u8, kind, "boolean")) return Field.boolean(label);
    if (std.mem.eql(u8, kind, "file")) return Field.file(label);
    if (std.mem.eql(u8, kind, "ref")) return Field.reference(label, entity orelse "");
    return Field.string(label);
}
